### DISCUSSION TITLE
Floris Blockage models

### DISCUSSION THREAD

**Discussion by Ooussamaa** [NatLabRockies/floris#615] [Q&A] [answered]
Hey guys, I have a specific question for you 
Now I can see that wake models exist in Floris and I know how to activate them, but blockage models I they exist but my question is: they are already coupled with blockage models so the AEP I calculate takes into account Wake and blockage effects at the same time or wake only ? and if it was the case, can you please push me in the right direction to find how to activate them coz I have searched in the litterature and I could not find my answer.
My purpose is: calculate the AEP for a wind farm by taking into account blockage and wake at the same time !

Thank you

**Comment by kilojoules** [comment]
I am also having trouble finding how to model blockage with floris...

**Reply by Ooussamaa** [reply]
Have you already managed to do so with pywake ?

**Reply by kilojoules** [reply]
Yeah https://topfarm.pages.windenergy.dtu.dk/PyWake/notebooks/EngineeringWindFarmModels.html#All2AllIterative

**Reply by Ooussamaa** [reply]
Thank you !

**Reply by Ooussamaa** [reply]
Hello Kilojoules,

I still facing some troubles in importing and modifying the blockage models on pywake 

May I have the python script you've developed to run your simulations  please !

Thank you

**Reply by kilojoules** [reply]
You can explore many of the options in pywake using this notebook: https://github.com/DTUWindEnergy/PyWake/blob/master/docs/notebooks/exercises/CombineModels.ipynb. 

Here is a simple example

<img width="843" alt="image" src="https://user-images.githubusercontent.com/8032238/230618069-cce0c95d-8426-4327-9731-31b980049d96.png">



```
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import time
from py_wake.examples.data.hornsrev1 import Hornsrev1Site, HornsrevV80
from py_wake.deficit_models import *
from py_wake.wind_farm_models import *
from py_wake.rotor_avg_models import *
from py_wake.superposition_models import *
from py_wake.deflection_models import *
from py_wake.turbulence_models import *
from py_wake.ground_models import *
from py_wake.deficit_models.utils import *

site = Hornsrev1Site()
windTurbines = HornsrevV80()
x,y = ([0, 500], [0, 0])

# windFarmModel autogenerated by dropdown boxes
t = time.time()
wfm = All2AllIterative(
    site,
    windTurbines,
    wake_deficitModel=BastankhahGaussianDeficit(
        k=0.0324555,
        ceps=0.2,
        use_effective_ws=False,
        rotorAvgModel=None,
        groundModel=None),
    superpositionModel=LinearSum(),
    blockage_deficitModel=VortexDipole(
        sct=1.0,
        limiter=1e-10,
        exclude_wake=True,
        superpositionModel=None,
        rotorAvgModel=None,
        groundModel=None,
        upstream_only=False),
    deflectionModel=None,
    turbulenceModel=None,
    convergence_tolerance=1e-06,
    initialize_with_PropagateDownwind=True,
    rotorAvgModel=None)
sim_res = wfm(x,y)
plt.figure(figsize=(12,8))
sim_res.flow_map(wd=270).plot_wake_map(cmap='viridis')
print (wfm)
print ("Computation time (AEP + flowmap):", time.time()-t)
plt.title('AEP: %.2fGWh'%(sim_res.aep().sum()))
plt.savefig('blockage_ex')
plt.clf()
```

![blockage_ex](https://user-images.githubusercontent.com/8032238/230615665-1b901ede-96d8-4dfc-aa2c-6921f22c716f.png)

**Reply by Ooussamaa** [reply]
Thanks !

**Reply by Ooussamaa** [reply]
Hello,
I am wondering if in Pywake we have the possibility to import a file for the layout we want to use as we can do in Floris with the format .CSV using pandas . Coz I want to modify the script that do the coupling of  wake and blockage for only two turbines and extend my study to a whole wind farm (if it is possible I wish if you can indicate me the right commands/ Modifications to do so)  or we are limited to the few wind farm layouts we have already in Pywake ?
Thanks a lot !

**Reply by kilojoules** [reply]
If you want to create more turbines, simply adjust the `x` and `y` lists to the corresponding x and y positions. If you are dealing with measured atmospheric data, this is covered in our documentation, e.g., https://topfarm.pages.windenergy.dtu.dk/PyWake/notebooks/Quickstart.html https://topfarm.pages.windenergy.dtu.dk/PyWake/notebooks/Site.html

I think we should move this discussion away from the floris repository. If you become stuck, you can send me an email. [This is my contact information](https://orbit.dtu.dk/en/persons/julian-quick). 

Cheers!

**Comment by rafmudaf** [comment]
@ebranlard see @kilojoules's question below (copied from https://github.com/NREL/floris/issues/616#issuecomment-1497382645). Are the blockage models you've described in the paper publicly available?

> This paper discusses some capabilities of modeling blockage in FLORIS: https://www.nrel.gov/docs/fy20osti/76855.pdf. It is not clear to me how these results can be reproduced. Is there an example available regarding how to model blockage in FLORIS?

**Comment by ebranlard** [comment]
Hi, 

The repository used for the FLORIS model with blockage effect presented in the paper (https://www.nrel.gov/docs/fy20osti/76855.pdf), contains the source code and some usage examples:
   https://github.com/ebranlard/wiz 

Unfortunatley, this repository was not updated for a while and it is therefore using an older version of FLORIS. 

I hope that helps,

Emmanuel

**Comment by Ooussamaa** [comment]
Hi @kilojoules,

I hope you are doing well.

I am simulating Wake and Blockage Coupling, as the Blockage effects are too small I am not able to visualize them clearly, can you please help me to do the reescaling of the velocity plot from [1m/s - 10m/s] to [0.5m/s - 1m/s], so induction zone will be clear enough on the figure.

Here is the script 

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import time
from py_wake.examples.data.hornsrev1 import Hornsrev1Site, HornsrevV80
from py_wake.deficit_models import *
from py_wake.wind_farm_models import *
from py_wake.rotor_avg_models import *
from py_wake.superposition_models import *
from py_wake.deflection_models import *
from py_wake.turbulence_models import *
from py_wake.ground_models import *
from py_wake.deficit_models.utils import *
from py_wake.examples.data.hornsrev1 import Hornsrev1Site, V80, wt_x, wt_y, wt16_x, wt16_y
from py_wake.examples.data.iea37 import IEA37Site, IEA37_WindTurbines
from py_wake.wind_turbines import WindTurbine, WindTurbines
from py_wake.examples.data.dtu10mw import DTU10MW

dtu10mw ='Mywindturbine'
site=IEA37Site()
windTurbines = DTU10MW()

x,y = ([565761.6,
566664.0,568428.4,570192.7,571095.2,571957.0,572859.5,573721.4,574623.8,575526.3,577290.6,565517.9,566420.4,578111.9,580819.3,565477.3,567282.3,571754.0,573625.3,575363.9,582583.6,583486.1,568169.1,569046.6,569949.1,574420.8,579835.6,581640.5,583445.5,584347.9,585250.4,586152.9,566298.6,569908.5,572772.3,575434.2,576185.1,579795.0,580697.5,565355.5,582502.4,583404.9,584307.3,585209.8,586112.3,588819.7,568965.4,565314.9,573437.1,577047.0,588779.1,589681.5,590584.0,567119.9,572534.7,575201.5,576103.9,577908.9,580616.3,582421.2,583323.7,585128.6,586056.1,587836.0,588785.9,589640.9,591445.9,567981.7,592348.3,570689.1,577868.3,578770.7,585990.5,586892.9,589600.3,590502.8,591405.3,577827.7,578730.1,579632.6,581437.5,582340.0,583242.5,584144.9,585949.9,586852.3,587754.8,588657.3,589559.7,590462.2,577787.1,578689.5,579592.0,585909.3,577746.4,578648.9,579551.4,580453.8,581356.3,582258.8,583161.3,584063.7,584966.2,585868.7,586771.1,588623.6,589478.5,586730.5,588535.5,582177.6,583080.0,583982.5,584885.0,587592.4],[6264687.0,6264165.9,6264616.7,6265067.4,6264546.3,6265518.1,6264997.1,6265968.8,6265447.8,6264926.8,6265377.5,6273643.8,6273122.8,6267842.1,6266278.9,6275136.6,6274094.5,6272982.2,6272014.8,6270898.0,6266729.7,6266208.6,6275109.6,6274545.3,6274024.2,6272911.8,6269785.6,6268743.5,6267701.4,6267180.4,
6266659.4,6266138.3,6277601.2,6275517.0,6275668.9,6274054.4,6273362.6,6271278.4,6270757.4,6279615.0,6269715.3,6269194.2,6268673.2,6268152.2,6267631.1,6266068.0,6277530.9,6281107.9,6276418.5,6274334.3,6267560.8,6267039.8,6266518.7,6280065.8,6276939.5,6276869.2,6276348.2,6275306.1,6273743.0,6272700.9,
6272179.9,6271137.8,6270573.4,6269574.7,6269055.0,6268532.6,6267490.5,6281037.5,6266969.5,6279474.4,6276798.9,6276277.9,6272109.5,6271588.5,6270025.4,
6269504.3,6268983.3,6278291.7,6277770.7,6277249.6,6276207.6,6275686.5,6275165.5,6274644.4,6273602.4,6273081.3,6272560.3,6272039.2,6271518.2,6270997.2,
6279784.5,6279263.5,6278742.4,6275095.2,6281277.3,6280756.3,6280235.2,6279714.2,6279193.2,6278672.1,6278151.1,6277630.0,6277109.0,6276588.0,6276066.9,
6275026.2,6274503.8,6277559.7,6276517.7,6281657.7,6281136.7,6280615.7,6280094.6,6278531.5])

t = time.time()
wfm = All2AllIterative(
    site,
    windTurbines,
    wake_deficitModel=NOJDeficit(
        k=0.1,
        rotorAvgModel=None,
        groundModel=None),
    superpositionModel=LinearSum(),

    blockage_deficitModel=VortexDipole(
        limiter=1e-10,
        sct = 3,
        exclude_wake=True,
        superpositionModel=None,
        rotorAvgModel=None,
        groundModel=None,
        upstream_only=False),
    deflectionModel=None,
    turbulenceModel=STF2017TurbulenceModel(),
    convergence_tolerance=1e-06,
    rotorAvgModel=None)
sim_res = wfm(x, y)
fig, ax = plt.subplots(figsize=(12, 8))
wake_map = sim_res.flow_map(wd=270).plot_wake_map(cmap='viridis')
plt.title('AEP: %.2fGWh'%(sim_res.aep().sum()))
plt.show()


Thank you by advance !
Best regards 
Oussama

**Reply by kilojoules** [reply, marked as answer]
Hi Oussama,

Please direct PyWake questions to our issue page here: [https://gitlab.windenergy.dtu.dk/TOPFARM/PyWake/-/issues](https://gitlab.windenergy.dtu.dk/TOPFARM/PyWake/-/issues)

Best,
Julian Quick

**Comment by christiannvaughn** [comment]
Hi @Ooussamaa, Have you tried finding your answer on PyWake’s GitLab page?
https://gitlab.windenergy.dtu.dk/TOPFARM/PyWake
GitLab doesn’t have the same forum features as GitHub. Alternatively, maybe @kilojoules can point you in the right direction.
